File: frontend/lib/api/document-qa.ts
const API_BASE_URL = 'http://localhost:8000'

// 會話管理 API
export interface SessionInfo {
  id: string
  name: string
  created_at: string
  document_count: number
}

export class SessionAPI {
  static async createSession(): Promise<{ session_id: string; message: string }> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/sessions`, {
      method: 'POST',
    })
    
    if (!response.ok) {
      throw new Error(`創建會話失敗: ${response.statusText}`)
    }
    
    return response.json()
  }
  
  static async listSessions(): Promise<SessionInfo[]> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/sessions`)
    
    if (!response.ok) {
      throw new Error(`獲取會話列表失敗: ${response.statusText}`)
    }
    
    return response.json()
  }
  
  static async addDocumentToSession(sessionId: string, docId: string): Promise<{ message: string }> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/sessions/${sessionId}/documents/${docId}`, {
      method: 'POST',
    })
    
    if (!response.ok) {
      throw new Error(`添加文檔失敗: ${response.statusText}`)
    }
    
    return response.json()
  }
  
  static async removeDocumentFromSession(sessionId: string, docId: string): Promise<{ message: string }> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/sessions/${sessionId}/documents/${docId}`, {
      method: 'DELETE',
    })
    
    if (!response.ok) {
      throw new Error(`移除文檔失敗: ${response.statusText}`)
    }
    
    return response.json()
  }
  
  static async getSessionDocuments(sessionId: string): Promise<DocumentInfo[]> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/sessions/${sessionId}/documents`)
    
    if (!response.ok) {
      throw new Error(`獲取會話文檔失敗: ${response.statusText}`)
    }
    
    return response.json()
  }
}

export interface DocumentInfo {
  id: string
  filename: string
  size: number
  content_type: string
  upload_time: string
  status: string
}

export interface SourceInfo {
  file_name: string
  page?: number
  chunk_id?: string
}

export interface SourceDocumentMetadata {
  filename: string
  chunks_count: number
  relevance_score: number
  expandable: boolean
}

export interface SourceDocument {
  snippet: string              // 相關片段 (150-200字)
  highlighted: string          // 高亮版本
  full_content: string         // 完整內容（供展開使用）
  metadata: SourceDocumentMetadata
}

export interface QuestionResponse {
  id: string
  question: string
  answer: string
  sources: SourceInfo[]        // 保留原有格式兼容性
  source_documents?: SourceDocument[]  // 新增優化的來源預覽
  timestamp: string
}

export interface SummaryResponse {
  summary: string
  key_points: string[]
  timestamp: string
}

export interface QuizQuestion {
  id: number
  question: string
  options: string[]
  correct_answer: number
}

export interface QuizResponse {
  questions: QuizQuestion[]
  timestamp: string
}

// 文檔問答 API
export class DocumentQAAPI {
  static async uploadDocument(file: File): Promise<DocumentInfo> {
    const formData = new FormData()
    formData.append('file', file)

    const response = await fetch(`${API_BASE_URL}/api/document-qa/upload`, {
      method: 'POST',
      body: formData,
    })

    if (!response.ok) {
      throw new Error(`上傳失敗: ${response.statusText}`)
    }

    return response.json()
  }

  static async listDocuments(): Promise<DocumentInfo[]> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/documents`)
    
    if (!response.ok) {
      throw new Error(`獲取文件列表失敗: ${response.statusText}`)
    }

    return response.json()
  }

  static async deleteDocument(docId: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/documents/${docId}`, {
      method: 'DELETE',
    })

    if (!response.ok) {
      throw new Error(`刪除文件失敗: ${response.statusText}`)
    }
  }

  static async askQuestion(question: string, documentIds: string[], sessionId?: string): Promise<QuestionResponse> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/question`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        question,
        document_ids: documentIds,
        session_id: sessionId,
      }),
    })

    if (!response.ok) {
      throw new Error(`問答失敗: ${response.statusText}`)
    }

    return response.json()
  }

  static async expandSourceContent(fileName: string, query?: string): Promise<{
    file_name: string
    content: string
    highlighted_content: string
    length: number
  }> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/expand-source`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        file_name: fileName,
        query: query || '',
      }),
    })

    if (!response.ok) {
      throw new Error(`展開內容失敗: ${response.statusText}`)
    }

    return response.json()
  }

  static async generateSummary(documentIds?: string[], sessionId?: string): Promise<SummaryResponse> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/summary`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        document_ids: documentIds,
        session_id: sessionId,
      }),
    })

    if (!response.ok) {
      throw new Error(`生成摘要失敗: ${response.statusText}`)
    }

    return response.json()
  }

  static async generateQuiz(documentIds?: string[], numQuestions: number = 5, sessionId?: string): Promise<QuizResponse> {
    const response = await fetch(`${API_BASE_URL}/api/document-qa/quiz`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        document_ids: documentIds,
        num_questions: numQuestions,
        session_id: sessionId,
      }),
    })

    if (!response.ok) {
      throw new Error(`生成測驗失敗: ${response.statusText}`)
    }

    return response.json()
  }
}

File: backend/app/models/document_qa.py
from pydantic import BaseModel
from typing import List, Optional
from datetime import datetime
from typing import Dict, Any

class DocumentUpload(BaseModel):
    """文件上傳請求模型"""
    filename: str
    content_type: str

class DocumentInfo(BaseModel):
    """文件信息模型"""
    id: str
    filename: str
    size: int
    content_type: str
    upload_time: datetime
    status: str  # "processing", "ready", "error"

class QuestionRequest(BaseModel):
    """問答請求模型"""
    question: str
    document_ids: Optional[List[str]] = None
    session_id: Optional[str] = None

class SourceInfo(BaseModel):
    """來源信息模型"""
    file_name: str
    page: Optional[int] = None
    chunk_id: Optional[str] = None

class SourceDocumentMetadata(BaseModel):
    """來源文檔元數據模型"""
    filename: str
    chunks_count: int
    relevance_score: float
    expandable: bool

class SourceDocument(BaseModel):
    """優化後的來源文檔模型"""
    snippet: str              # 相關片段 (150-200字)
    highlighted: str          # 高亮版本
    full_content: str         # 完整內容（供展開使用）
    metadata: SourceDocumentMetadata

class QuestionResponse(BaseModel):
    """問答回應模型"""
    id: str
    question: str
    answer: str
    sources: List[SourceInfo]  # 保留原有格式兼容性
    source_documents: Optional[List[SourceDocument]] = []  # 新增優化的來源預覽
    timestamp: datetime

class SummaryRequest(BaseModel):
    """摘要請求模型"""
    document_ids: Optional[List[str]] = None
    session_id: Optional[str] = None

class SummaryResponse(BaseModel):
    """摘要回應模型"""
    summary: str
    key_points: List[str]
    timestamp: datetime

class QuizRequest(BaseModel):
    """測驗請求模型"""
    document_ids: Optional[List[str]] = None
    session_id: Optional[str] = None
    num_questions: int = 5

class QuizQuestion(BaseModel):
    """測驗題目模型"""
    id: int
    question: str
    options: List[str]
    correct_answer: int

class QuizResponse(BaseModel):
    """測驗回應模型"""
    questions: List[QuizQuestion]
    timestamp: datetime